src/lem/constrainer.rs

```diff
@@ -0,0 +1,344 @@
+    /// constrained later, using hash maps to manage viariables and pointers in
+    /// a way we can reference allocated variables that were previously created.
+    ///
+    /// Notably, we use a variable `branch_path_info: Option<Boolean>` to encode
+    /// a three-way information:
+    ///
+    /// * If it's `None`, it means that no logical branches have happened in the
+    /// LEM so far, meaning that the evaluation algorithm must have gone through
+    /// that operation. In this case, we use regular equality enforcements
+    ///
+    /// * If it's `Some(concrete_path)`, it means that we're on a logical LEM
+    /// branch, which might be *virtual* or *concrete* depending on the witness.
+    /// A virtual path is one that wasn't taken during evaluation and thus its
+    /// witness pointers and variables weren't bound. A concrete path means that
+    /// evaluation took that path and the witness data should be complete. For
+    /// virtual paths we need to create dummy bindings and relax the enforcements
+    /// with implications whose premises are false. So, in the end, we use
+    /// implications on both virtual and concrete paths to make sure that the
+    /// circuit structure is always the same, independently of the witness. The
+    /// premise is precicely `concrete_path`.
+    pub fn constrain<F: LurkField, CS: ConstraintSystem<F>>(
+        &self,
+        cs: &mut CS,
+        store: &mut Store<F>,
+        witness: &Witness<F>,
+    ) -> Result<(), String> {
+        let mut alloc_ptrs: HashMap<&String, AllocatedPtr<F>> = HashMap::default();
+
+        // allocate first input
+        self.allocate_and_inputize_input(cs, store, witness, &mut alloc_ptrs, 0)?;
+
+        // allocate second input
+        self.allocate_and_inputize_input(cs, store, witness, &mut alloc_ptrs, 1)?;
+
+        // allocate third input
+        self.allocate_and_inputize_input(cs, store, witness, &mut alloc_ptrs, 2)?;
+
+        let mut num_inputized_outputs = 0;
+
+        let mut alloc_manager = AllocationManager::default();
+
+        let mut stack = vec![(&self.lem_op, None::<Boolean>, String::new())];
+        while let Some((op, branch_path_info, path)) = stack.pop() {
+            match op {
+                LEMOP::MkNull(tgt, tag) => {
+                    if alloc_ptrs.contains_key(tgt.name()) {
+                        return Err(format!("{} already allocated", tgt.name()));
+                    };
+                    let z_ptr = {
+                        if Self::on_concrete_path(&branch_path_info)? {
+                            let Some(ptr) = witness.ptrs.get(tgt.name()) else {
+                                return Err(format!("Couldn't retrieve witness {}", tgt.name()));
+                            };
+                            store.hydrate_ptr(ptr)?
+                        } else {
+                            ZPtr::dummy()
+                        }
+                    };
```

Comment:
```suggestion
                    let z_ptr = if Self::on_concrete_path(&branch_path_info)? {
                        let ptr = witness
                            .ptrs
                            .get(tgt.name())
                            .ok_or_else(|| anyhow!("Couldn't retrieve witness {}", tgt.name()))?;
                        store.hydrate_ptr(ptr)?
                    } else {
                        ZPtr::dummy()
                    };
```

Reply:
Can't apply suggestion due to conflict with previous suggestion